### PR TITLE
create pattern construct for a webapp extension

### PR DESCRIPTION
**Pull request by thrau**
## Motivation

Webapp extensions such as:
* https://github.com/localstack/localstack-ngrok-tunnels
* https://github.com/thrau/localstack-aws-log-viewer
* https://github.com/localstack/localstack-extension-event-studio

Follow a common pattern. They
* expose routes through subdomains and submounts (`http://my-extension.localhost.localstack.cloud:4566`, and `http://localhost.localstack.cloud:4566/_extension/my-extension`)
* Enable logging for their root module `my_extension`
* Serve jinja2 templates
* Serve static files

Their file layout will typically look like this:
```
my_extension
├── extension.py
├── __init__.py
├── static              <-- make sure static resources get packaged!
│   ├── __init__.py
│   ├── favicon.ico
│   └── style.css
└── templates            <-- jinja2 templates
   └── index.html
```

This PR introduces this pattern as a class that can be extended to quickly develop extensions like this.

Examples of how existing extensions have been simplified can be found in:
* https://github.com/thrau/localstack-aws-log-viewer/pull/1
* https://github.com/localstack/localstack-ngrok-tunnels/commit/1838953260e44ca31046928f9576c7902abc17d1
* https://github.com/localstack/localstack-extension-event-studio

Requires [Rolo v0.5.0](https://github.com/localstack/rolo/releases/tag/v0.5.0)

## Changes

* Introduce `localstack.extension.patterns`
* Add a high-level pattern `WebAppExtension`

## Testing

If you want to test this you can:

* download and checkout https://github.com/thrau/localstack-aws-log-viewer/pull/1 (or just build your own extension if you want)
  * run `make install`
  * run `localstack extensions dev enable .`
* checkout this branch
* in localstack-ext, run `EXTENSION_DEV_MODE=1 python -m localstack.dev.run -e LOCALSTACK_API_KEY=test -e EXTENSION_DEV_MODE=1`

## TODO

What's left to do:

- [x] Wait for rolo dependency to be updated v0.5.0

### CODE REVIEW

File: localstack/extensions/patterns/webapp.py
Comment (joe4dev) [inline localstack/extensions/patterns/webapp.py:79]: What happens upon naming clashes? the last registered route wins 🤔

File: localstack/extensions/patterns/webapp.py
Comment (joe4dev) [inline localstack/extensions/patterns/webapp.py:17]: might be worth commenting that `jinja2` is a peer dependency that needs to be installed through the extension for templating

File: localstack/extensions/patterns/webapp.py
Comment (joe4dev) [inline localstack/extensions/patterns/webapp.py:180]: Let's hope this jinja2 API never changes because the peer dependency in the extension needs to package a compatible version 🤞

File: localstack/extensions/patterns/container.py
Comment (joe4dev) [inline localstack/extensions/patterns/container.py:1]: Are these empty files intentional?

File: localstack/extensions/patterns/webapp.py
Comment (joe4dev) [inline localstack/extensions/patterns/webapp.py:235]: nice catch 💯 
tested ✔️

**Review by joe4dev — COMMENTED** [review]
Looking at the extension examples, it makes the API so much easier to use. Auto-CORS config is a nice extra as well. 🙌 

I tested these changes with the https://github.com/localstack/localstack-extension-event-studio and it works like a charm. 👍 

I'm not so familiar with our routing internals and would suggest someone else signing it off, for example @alexrashed or @dfangl.

File: localstack/extensions/patterns/webapp.py
Comment (alexrashed) [inline localstack/extensions/patterns/webapp.py:59]: nit: I guess this already evolved and is used for existing extensions, which would make it hard to change, but I would suggest to use the plural for API paths that contain lists:
```suggestion
    * Submount: https://localhost.localstack.cloud:4566/_extensions/my-extension
```

And I guess it was already discussed in the past if the `_extension` endpoint should be a top-level entry / first class citizen in LocalStack? For example, CloudPods live in `/_localstack/pods`. But extensions can also be totally unrelated to LocalStack itself (like `mailhog`), where it might even be confusing...

File: localstack/extensions/patterns/webapp.py
Comment (alexrashed) [inline localstack/extensions/patterns/webapp.py:60]: nit / q: If there are a _lot_ of extensions this might become problematic because this could overlap with other domain-based routes. Would it make sense to have them combined in a defined subcomain like `my-extension.extensions.localhost.localstack.cloud`?

File: localstack/extensions/patterns/container.py
Comment (alexrashed) [reply localstack/extensions/patterns/container.py:1]: +1 for removing this file before merging. :)

**Review by alexrashed — APPROVED** [review]
Really nice abstraction, which saves a lot of repetitive code among the different extension implementations, but also helps tackling edge cases in a generic way (f.e. the 301 redirect for relative imports)! 🚀 
I think this can be merged as is (after removing the `container.py`). 🥳

File: localstack/extensions/patterns/webapp.py
Comment (thrau) [reply localstack/extensions/patterns/webapp.py:79]: True. nothing we can do about that right now unfortunately. Added a comment though

File: localstack/extensions/patterns/webapp.py
Comment (thrau) [reply localstack/extensions/patterns/webapp.py:180]: for now it should be OK since we have a pin on jinja through moto. in the worst case people can override the method in their extension.

but the issue of dependencies in extensions (and potential conflicts) is something we definitely need to tackle at some point.

File: localstack/extensions/patterns/container.py
Comment (thrau) [reply localstack/extensions/patterns/container.py:1]: i gave away the other patterns i'm playing with :D

File: localstack/extensions/patterns/webapp.py
Comment (thrau) [reply localstack/extensions/patterns/webapp.py:59]: well it hasn't really been discussed directly, i'm making a "product owner" decision here to introduce it. i believe having a well-known namespace for extensions is important to have since it allows us to reason over the requests. moreover, it should be distinct from `/_localstack` for the reason you correctly identified. we should port our existing extensions to that convention.

i prefer singular for the following (completely arbitrary and personal taste reason which you are welcome to disagree with :D). we have `/_localstack/extensions` which indicates that you can _manage_ one or more extensions. whereas the singular `/_extension/<something>` indicates that you're talking to an individual and specific extension.

Comment (alexrashed) [reply localstack/extensions/patterns/webapp.py:59]: I feel like it's a best practice when it comes to HTTP / RESTful API design to use the plural for path segments that in turn contains a collection of sub-resources (`/_extensions` is a collection resource whereas the specific extension route would expose a single resource). I don't see a reason to use the `/_extension(s)` path in the future to show metadata on the _collection_ of extensions registered in this "namespace", for example a `GET /_extensions` listing the different sub-routes for `WebAppExtensions`...
But again, just a nitpick 😛

Comment (thrau) [reply localstack/extensions/patterns/webapp.py:59]: All fair points. This is not a REST API though :)

File: localstack/extensions/patterns/webapp.py
Comment (thrau) [reply localstack/extensions/patterns/webapp.py:60]: definitely a good point that we can explore! for now i'll keep this so we don't need to create a new wildcard certificate. for `*.extension....`

**Comment by thrau** [comment]
marking this as patch since it is purely additive and affects no existing localstack control paths

**Pull request merged**
